drivers/pcie/pcie.c

```diff
@@ -78,3 +83,21 @@ u32_t pcie_get_iobar(pcie_bdf_t bdf, unsigned int index)
 {
 	return pcie_get_bar(bdf, index, true);
 }
+
+bool pcie_irq_enable(pcie_bdf_t bdf, unsigned int irq)
```

Comment:
Design-wise, seems like this is mixing two different tasks: enabling MSI interrupt targetting on the device and masking the interrupt locally in the APIC (which is what irq_en/disable() does).

Do we need that?  Seems like it would be clearer if MSI setup (which is obviously device-dependent) was visible in the driver as a separate step from cross-architecture-portable IRQ masking.

Reply:
Unfortunately, for the moment, configuring MSI and enabling the associated IRQ are indivisible operations.  There are two factors that contribute to this:

1. Once MSIs are enabled at the endpoint, the standard says they can NOT be disabled.

2. MSIs are delivered directly to the local APIC of the target CPU, and the local APICs do not have a mechanism to disable individual vectors (barring, e.g., the timer, which is local to the APIC itself).  Local APICs can block classes of IRQs using task priority, but that's not suitable.

So configuring MSI is the same as enabling the IRQ.

This means, of course, that PCI(e)-aware drivers must be written with the understanding that they can't synchronize by blocking their device's IRQ.

Reply:
Yeah, good point.  Though I still think this argues for a framework that merges the generic API with some MSI intelligence on the backend, but that can come later once we understand the problem and portability needs.

Reply:
Agreed.  This is suboptimal.